### ISSUE TITLE
Hello world failed after upgrading customized Trusted OS to optee 3.3.0

### ISSUE THREAD

**Issue by vikaspot** [OP-TEE/optee_os#2860]

**Comment by jbech-linaro** [comment]
Why do you mean by customized Trusted OS?  I'm running latest (`OP-TEE version: 3.4.0-117-gdba5a1ea #3 Tue Mar  5 12:28:45 UTC 2019 arm`)

NW console
```bash
# optee_example_hello_world 
Invoking TA to increment 42
TA incremented value to 43
```
SW console
```bash
D/TC:? 0 tee_ta_init_pseudo_ta_session:280 Lookup pseudo TA 8aaaf200-2450-11e4-abe2-0002a5d5c51b
D/TC:? 0 load_elf:833 Lookup user TA ELF 8aaaf200-2450-11e4-abe2-0002a5d5c51b (Secure Storage TA)
D/TC:? 0 load_elf:833 Lookup user TA ELF 8aaaf200-2450-11e4-abe2-0002a5d5c51b (REE [buffered])
D/TC:? 0 load_elf_from_store:801 ELF load address 0x103000
D/TC:? 0 tee_ta_init_user_ta_session:1023 Processing relocations in 8aaaf200-2450-11e4-abe2-0002a5d5c51b
D/TC:? 0 tee_ta_close_session:398 csess 0xe171b50 id 1
D/TC:? 0 tee_ta_close_session:417 Destroy session
D/TC:? 0 tee_ta_close_session:443 Destroy TA ctx
```

I.e., seems to be working as it should.

**Comment by vikaspot** [comment]
I have modified my Optee_os component from one Version 3.0.0 to 3.3.0 
But hello_world Sample app is not working .
here is log 
 hello_world

D/HC:? 0 tee_ta_init_pseudo_ta_session:273 Lookup pseudo TA 8aaaf200-2450-11e4-abe2-0002a5d5c51b
D/HC:? 0 load_elf:842 Lookup user TA ELF 8aaaf200-2450-11e4-abe2-0002a5d5c51b (REE)
D/HC:? 0 load_elf_from_store:810 ELF load address 0x80005000
D/HC:? 0 tee_ta_init_user_ta_session:1029 Processing relocations in 8aaaf200-2450-11e4-abe2-0002a5d5c51b
E/HC:? 0 tee_ta_init_user_ta_session:1052 Invalid TA flag(s) 0x3
D/HC:? 0 tee_ta_open_session:540 init session failed 0xffff0005
hello_world: TEEC_Opensession failed with code 0xffff0005 origin 0x3

Pls help !! What should b reason for session not to get succeed ?

**Comment by vikaspot** [comment]
Customize means .. : in My OS Optee_os is part where i am using it along with optee_client and vendor specific ATF and my customized loader !!

**Comment by jbech-linaro** [comment]
How have you debugged it and what are your conclusions after doing the debugging so far?

**Comment by jforissier** [comment]
> E/HC:? 0 tee_ta_init_user_ta_session:1052 Invalid TA flag(s) 0x3

**Comment by vikaspot** [comment]
No idea . Still i am debugging . As per search and debugging .. I was suspecting .. any trusted application is not allowed to run is some kind of security violation. For proper context/session is not created properly so the parameters are not correct .. I am not much expert . But ok to related internal .. .. any instance for ta app is not allowed .. ?

**Comment by jforissier** [comment]
@vikaspot are you trying to run an "old" TA (compiled with dev kit from OP-TEE 3.0.0) with a newer version of OP-TEE (3.3.0)?
I would not be surprised if commit https://github.com/OP-TEE/optee_os/commit/387b0ee3 had broken backward compatibility :(

**Comment by jbech-linaro** [comment]
First I think something has went wrong when you've done the upgrade from 3.0.0 to 3.3.0. I'm saying that since the hello world TA is running in our automated testing environment and I cannot recall that I've ever seen the hello world TA fail there. However, since this seems to be a modified OP-TEE version you're running, it might be that it's something wrong being introduced in the fork (I see `HC` in your debug prints, so this is for sure not running upstream code).

Secondly, if you know how to use gdb, use that. If not, then start adding debug prints (`DMSG`) around the code where it's failing, you have both the function names and the line numbers in your log.

Another thing that you could try (you probably already have tried this) is to to wipe all built files to ensure that there are no stale files, typically:
```bash
$ cd <optee-project>/<optee_xyz> && git clean -xdf
$ rm -rf <optee_project>/out
$ rm -rf <optee_project>/out-br
```
Then recompile it all and test again.

> any instance for ta app is not allowed .. ?

No, they are always allowed to run as long as the signature has been verified OK. The hello world TA is really simplistic and contains no magic at all.

**Comment by vikaspot** [comment]
Thanks for advise,
any instance for ta app is not allowed .. ?
VP>> Need to check and confirm

**Comment by jbech-linaro** [comment]
FYI, I've just sent a documentation pull request (https://github.com/OP-TEE/optee_docs/pull/14) that describes how to debug TEE core with GDB using QEMU, not sure if that is something useful for you or not.

**Comment by jbech-linaro** [comment]
It's merged now, you will find the rendered version [here](https://optee.readthedocs.io/building/devices/qemu.html#gdb-secure-world).

**Comment by vikaspot** [comment]
Please correct me if I am wrong ..
I have used ,optee_os,optee_client, sample app all 3 components as 3.3.0 and built .

But still I faced same errors .. .
I want to know more ,how can I allign all optee related components in same version.
Your observations regarding (HC) in my debug prints what does it indicates ? In 3.3.0 version what is expected ?

**Comment by jforissier** [comment]
@vikaspot this is what I mean: https://github.com/OP-TEE/optee_os/pull/2861

**Comment by vikaspot** [comment]
Thanks for help ..
@Jérôme
You guys are true . I had issue with reference of dev_kit.mk  it was taking from wrong place ... 
I have corrected all dev_kit and lib dependency all conf.mk ..etc dependency.
Now all components requirement to build trust App is correct and takeing reference from optee_os/out/export-ta-mediatek/xyz 


But still I think there is some more dependency I need to solve.

Can you pls suggest .  For any ta app to be built what's dependency lib/compilation files/..etc is required

**Comment by jforissier** [comment]
> Can you pls suggest . For any ta app to be built what's dependency lib/compilation files/..etc is required

All you need to build a TA is in `export-ta_arm32` or `export-ta_arm64`.

**Comment by vikaspot** [comment]
Ok .  I'll plan to recheck again . Thanks

**Comment by vikaspot** [comment]
I am using hello_world app from 
https://github.com/linaro-swg/optee_examples/tree/master/hello_world
 relased version of 3.3.0 
Is it correct  ?

if not please suggest some other link to use Hello_world app.

**Comment by jbech-linaro** [comment]
That's correct, that's the same some one as I used initially in this thread.

**Comment by vikaspot** [comment]
yes . .. Its working now Thanks a lot for quick and meaningful guide lines... which helped me . to get it solved,
Actually is "HC"is out  prefix Extension 
issue was optee_client,lates  optee_os/out/.../export-ta_arm64  and one flag which i  should customise 
in core/arch/arm/arm.mk 
CFG_SCTLR_ALIGNMENT_CHECK ?= n  (y => n) required for our platform
==================
hello_world
D/HC:? 0 tee_ta_init_pseudo_ta_session:273 Lookup pseudo TA 8aaaf200-2450-11e4-abe2-0002a5d5c51b
D/HC:? 0 load_elf:842 Lookup user TA ELF 8aaaf200-2450-11e4-abe2-0002a5d5c51b (REE)
D/HC:? 0 load_elf_from_store:810 ELF load address 0x80005000
D/HC:? 0 tee_ta_init_user_ta_session:1029 Processing relocations in 8aaaf200-2450-11e4-abe2-0002a5d5c51b
Invoking TA to increment 42
TA incremented value to 43
D/HC:? 0 tee_ta_close_session:380 tee_ta_close_session(0x45067d80)
D/HC:? 0 tee_ta_close_session:399 Destroy session
D/HC:? 0 tee_ta_close_session:425 Destroy TA ctx
root@richmond-9951-a0:/mnt/data/scr# hello_world
D/HC:? 0 tee_ta_init_pseudo_ta_session:273 Lookup pseudo TA 8aaaf200-2450-11e4-abe2-0002a5d5c51b
D/HC:? 0 load_elf:842 Lookup user TA ELF 8aaaf200-2450-11e4-abe2-0002a5d5c51b (REE)
D/HC:? 0 load_elf_from_store:810 ELF load address 0x80005000
D/HC:? 0 tee_ta_init_user_ta_session:1029 Processing relocations in 8aaaf200-2450-11e4-abe2-0002a5d5c51b
Invoking TA to increment 42
TA incremented value to 43
D/HC:? 0 tee_ta_close_session:380 tee_ta_close_session(0x450678a0)
D/HC:? 0 tee_ta_close_session:399 Destroy session
D/HC:? 0 tee_ta_close_session:425 Destroy TA ctx

          Thanks !! stay connected .. i'll contribute more in future,

**Comment by etienne-lms** [comment]
@vikaspot, great it works! envoy.
As for `CFG_SCTLR_ALIGNMENT_CHECK`, note you don't need to patch optee-os sources. You likely have a build script somewhere that runs something like `make PLATFORM=xxx ` to build `optee_os`. You can append `CFG_xxx=xxx` directives as extra arguments for OP-TEE core configuration, i.e.:
```
make PLATFORM=xxx CFG_SCTLR_ALIGNMENT_CHECK=n
```
And my apologies for the backward compat issue, looks like my contribution.


@all,
This Issue and this [thread](https://github.com/OP-TEE/optee_os/pull/2801#issuecomment-470438806) involves both TA versioning. The later proposes to solve it using a specific tag in the TA header that differentiate from a valid address. Maybe a kind of version tag to track TA compat capabilities. TA/core interface looks pretty stable, but at least these 2 raise versioning considerations.

**Comment by Sourabhharman** [comment]
Hi,
I faced the similar problem. But I tried this with android. I updated my OPTEE from 3.0.0 to 3.3.0.
What I did is, I built the OPTEE export-ta_arm64 components with ANDROID toolchain. I build my hello world TA's with 3.3.0 export-ta_arm64 components on ANDROID. 
I am getting this error:
_D/HC:? 0 tee_ta_init_pseudo_ta_session:273 Lookup pseudo TA 8aaaf200-2450-11e4-abe2-0002a5d5c51b
D/HC:? 0 load_elf:843 Lookup user TA ELF 8aaaf200-2450-11e4-abe2-0002a5d5c51b (REE)
D/HC:? 0 load_elf_from_store:811 ELF load address 0x40005000
D/HC:? 0 tee_ta_init_user_ta_session:1030 Processing relocations in 8aaaf200-2450-11e4-abe2-0002a5d5c51b
D/HC:? 0 tee_ta_close_session:380 tee_ta_close_session(0xbd0678a0)
D/HC:? 0 tee_ta_close_session:399 Destroy session
D/HC:? 0 tee_ta_close_session:425 Destroy TA ctx
E/HC:? 0 tee_ta_open_session:588 Failed. Return error 0x40004d40
hello_world: TEEC_Opensession failed with code 0x40004d40 origin 0x4_

The return error is not seems to be a GP error. I debug the hartos after that. I found that thread_enter_user_mode() is generating the error.

Is there anything wrong with the TA building?

FYI, I run the same set of OPTEE 3.3.0 image in linux, with the hello_world app running fine,
